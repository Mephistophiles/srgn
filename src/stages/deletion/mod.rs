use super::{tooling::StageResult, Stage};
use regex::Regex;

/// Deletes all matches of a given regex.
#[derive(Debug, Clone)]
#[allow(clippy::module_name_repetitions)]
pub struct DeletionStage {
    pattern: Regex,
}

impl Stage for DeletionStage {
    fn substitute(&self, input: &str) -> StageResult {
        Ok(self.pattern.replace_all(input, "").to_string().into())
    }
}

impl DeletionStage {
    /// Create a new [`DeletionStage`].
    ///
    /// # Arguments
    ///
    /// * `pattern`: The regex to use for deletion.
    #[must_use]
    pub fn new(pattern: Regex) -> Self {
        Self { pattern }
    }
}
