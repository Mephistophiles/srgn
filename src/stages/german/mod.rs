mod driver;
mod machine;
mod words;

// Re-export symbols.
#[allow(clippy::module_name_repetitions)]
pub use driver::GermanStage;
use words::{LetterCasing, SpecialCharacter, Umlaut, Word};
