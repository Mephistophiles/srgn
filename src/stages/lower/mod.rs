use crate::scoped::Scoped;

use super::Stage;

/// Renders in lowercase.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
#[allow(clippy::module_name_repetitions)]
pub struct LowerStage {}

impl Scoped for LowerStage {}

impl Stage for LowerStage {
    fn substitute(&self, input: &str) -> String {
        input.to_lowercase()
    }
}

#[cfg(test)]
mod tests {
    use rstest::rstest;

    use super::*;

    #[rstest]
    // German
    #[case("A", "a")]
    #[case("a", "a")]
    #[case("Ä", "ä")]
    #[case("ä", "ä")]
    #[case("Ö", "ö")]
    #[case("ö", "ö")]
    #[case("Ü", "ü")]
    #[case("ü", "ü")]
    #[case("ẞ", "ß")]
    #[case("ß", "ß")]
    #[case("AaÄäÖöÜüẞß!", "aaääööüüßß!")]
    #[case("SS", "ss")]
    //
    // Chinese
    #[case("你好!", "你好!")]
    //
    // Japanese
    #[case("こんにちは!", "こんにちは!")]
    //
    // Korean
    #[case("안녕하세요!", "안녕하세요!")]
    //
    // Russian
    #[case("ПРИВЕТ!", "привет!")]
    //
    // Emojis
    #[case("👋\0", "👋\0")]
    fn substitute(#[case] input: &str, #[case] expected: &str) {
        assert_eq!(LowerStage {}.substitute(input), expected);
    }
}
