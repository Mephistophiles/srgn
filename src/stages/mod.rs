#[cfg(feature = "de")]
mod german;
#[cfg(feature = "symbols")]
mod symbols;
/// Tooling (types, traits, ...) around stages.
pub mod tooling;

pub use german::GermanStage;
pub use symbols::SymbolsStage;
pub use tooling::Stage;
