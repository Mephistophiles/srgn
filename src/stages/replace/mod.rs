use log::info;
use unescape::unescape;

use super::Stage;

/// Replaces input with a fixed string.
///
/// ## Example: replacing invalid characters in identifiers
///
/// ```rust,ignore
/// use betterletters::RegexPattern;
/// use betterletters::stages::{Stage, ReplacementStage};
/// use betterletters::scoping::{ScopedViewBuilder, regex::Regex};
///
/// let stage = ReplacementStage::new("_".to_string());
/// let scoper = Regex::new(RegexPattern::new(r"[^a-zA-Z0-9]+").unwrap());
/// let mut view = ScopedViewBuilder::new("hyphenated-variable-name").explode_from_scoper(
///     &scoper
/// ).build();
///
/// assert_eq!(
///    stage.map(&mut view).to_string(),
///   "hyphenated_variable_name"
/// );
/// ```
///
/// ## Example: replace emojis
///
/// ```rust,ignore
/// use betterletters::RegexPattern;
/// use betterletters::stages::{Stage, ReplacementStage};
/// use betterletters::scoping::{ScopedViewBuilder, regex::Regex};
///
/// let stage = ReplacementStage::new(":(".to_string());
/// // A Unicode character class category. See also
/// // https://github.com/rust-lang/regex/blob/061ee815ef2c44101dba7b0b124600fcb03c1912/UNICODE.md#rl12-properties
/// let scoper = Regex::new(RegexPattern::new(r"\p{Emoji}").unwrap());
/// let mut view = ScopedViewBuilder::new("Party! 😁 💃 🎉 🥳 So much fun! ╰(°▽°)╯").explode_from_scoper(
///     &scoper
/// ).build();
///
/// assert_eq!(
///    stage.map(&mut view).to_string(),
///    // Party is over, sorry ¯\_(ツ)_/¯
///   "Party! :( :( :( :( So much fun! ╰(°▽°)╯"
/// );
/// ```
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct ReplacementStage {
    replacement: String,
}

impl TryFrom<String> for ReplacementStage {
    type Error = String;

    fn try_from(replacement: String) -> Result<Self, Self::Error> {
        let unescaped =
            unescape(&replacement).ok_or("Cannot unescape sequences in replacement".to_string())?;
        Ok(Self {
            replacement: unescaped,
        })
    }
}

impl Stage for ReplacementStage {
    fn process(&self, input: &str) -> String {
        info!("Substituting '{}' with '{}'", input, self.replacement);
        self.replacement.clone()
    }
}
