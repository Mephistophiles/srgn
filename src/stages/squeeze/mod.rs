use super::Stage;
use crate::{
    scoped::{ScopeStatus, Scoped},
    scoping::ScopedView,
};
use log::{debug, trace};
use regex::Regex;

/// Squeezes all consecutive matched scopes into a single occurrence.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
#[allow(clippy::module_name_repetitions)]
pub struct SqueezeStage {}

impl Scoped for SqueezeStage {}

impl Stage for SqueezeStage {
    fn substitute(&self, view: &mut ScopedView) {
        debug!("Squeezing...");
        let v = view.into_inner_mut();

        let mut prev_was_in = false;
        v.retain(|scope| {
            let keep = !(prev_was_in && matches!(scope, ScopeStatus::In(_)));
            prev_was_in = matches!(scope, ScopeStatus::In(_));
            trace!("keep: {}, scope: {:?}", keep, scope);
            keep
        });

        debug!("Squeezed: {:?}", v);

        // let mut previous = None;
        // for scope in v {
        //     if let ScopeStatus::In(_) = scope {
        //         if let Some(ScopeStatus::In(_)) = previous {
        //             continue;
        //         }
        //     }
        // }

        //     out.push_str((&scope).into());
        //     previous = Some(scope);
        // }

        // out
    }
}

#[cfg(test)]
mod tests {
    use rstest::rstest;

    use super::*;

    #[rstest]
    // Pattern only
    #[case("a", "a", "a")]
    #[case("aa", "a", "a")]
    #[case("aaa", "a", "a")]
    //
    // Pattern once; nothing to squeeze
    #[case("aba", "a", "aba")]
    #[case("bab", "a", "bab")]
    #[case("babab", "a", "babab")]
    #[case("ababa", "a", "ababa")]
    //
    // Squeezes only the pattern, no other repetitions
    #[case("aaabbb", "a", "abbb")]
    //
    // Squeezes start
    #[case("aab", "a", "ab")]
    //
    // Squeezes middle
    #[case("baab", "a", "bab")]
    //
    // Squeezes end
    #[case("abaa", "a", "aba")]
    //
    // Squeezes as soon as pattern occurs at least twice
    #[case("a", "ab", "a")]
    #[case("ab", "ab", "ab")]
    #[case("aba", "ab", "aba")]
    #[case("abab", "ab", "ab")]
    #[case("ababa", "ab", "aba")]
    #[case("ababab", "ab", "ab")]
    //
    // Squeezes nothing if pattern not present
    #[case("", "b", "")]
    #[case("a", "b", "a")]
    #[case("aa", "b", "aa")]
    #[case("aaa", "b", "aaa")]
    //
    // Deals with character classes (space)
    #[case("Hello World", r"\s", "Hello World")]
    #[case("Hello  World", r"\s", "Hello World")]
    #[case("Hello       World", r"\s", "Hello World")]
    #[case("Hello\tWorld", r"\t", "Hello\tWorld")]
    #[case("Hello\t\tWorld", r"\t", "Hello\tWorld")]
    //
    // Deals with character classes (inverted space)
    #[case("Hello World", r"\S", "H W")]
    #[case("Hello\t\tWorld", r"\S", "H\t\tW")]
    //
    // Deals with overlapping matches; behavior of `regex` crate
    #[case("abab", r"aba", "abab")]
    #[case("ababa", r"aba", "ababa")]
    #[case("ababab", r"aba", "ababab")]
    #[case("abababa", r"aba", "abababa")]
    //
    #[case("aba", r"aba", "aba")]
    #[case("abaaba", r"aba", "aba")]
    //
    // Turns greedy quantifiers into non-greedy ones automatically
    #[case("ab", r"\s+", "ab")]
    #[case("a b", r"\s+", "a b")]
    #[case("a\t\tb", r"\s+", "a\tb")]
    //
    // Turns greedy quantifiers into non-greedy ones automatically, even if user
    // specified themselves (extra option ignored)
    #[case("ab", r"(?U)\s+", "ab")]
    #[case("a b", r"(?U)\s+", "a b")]
    #[case("a\t\tb", r"(?U)\s+", "a\tb")]
    //
    // Inversion works (if user specified non-greedy manually, it becomes greedy). `+`
    // and `*` will make the concept of 'consecutive matches' meaningless!
    #[case("ab", r"\s+?", "ab")]
    #[case("a b", r"\s+?", "a b")]
    #[case("a\t\tb", r"\s+?", "a\t\tb")]
    #[case("a\t\t\t\tb", r"\s+?", "a\t\t\t\tb")]
    //
    // Deals with more complex patterns
    #[case("ab", "", "ab")] // Matches nothing
    //
    #[case("ab", r"[ab]", "a")]
    #[case("ab", r"[ab]+", "a")]
    #[case("ab", r"[ab]+?", "ab")]
    //
    #[case("abab", r"\D", "a")]
    //
    #[case("abab", r"(ab){2}", "abab")]
    #[case("ababa", r"(ab){2}", "ababa")]
    #[case("ababab", r"(ab){2}", "ababab")]
    #[case("abababa", r"(ab){2}", "abababa")]
    #[case("abababab", r"(ab){2}", "abab")]
    #[case("ababababa", r"(ab){2}", "ababa")]
    #[case("ababababab", r"(ab){2}", "ababab")]
    #[case("abababababab", r"(ab){2}", "abab")]
    //
    #[case("Anything whatsoever gets rEkT", r".", "A")]
    #[case(
        "Anything whatsoever gets rEkT",
        r".*", // Greediness inverted
        "Anything whatsoever gets rEkT"
    )]
    //
    // Deals with Unicode shenanigans
    #[case("ðŸ˜ŽðŸ˜Ž", r"ðŸ˜Ž", "ðŸ˜Ž")]
    #[case("\0ðŸ˜Ž\0ðŸ˜Ž\0", r"ðŸ˜Ž", "\0ðŸ˜Ž\0ðŸ˜Ž\0")]
    //
    #[case("ä½ ä½ å¥½", r"ä½ ", "ä½ å¥½")]
    //
    // Longer ("integration") tests; things that come up in the wild
    #[case(
        " dirty Strings  \t with  \t\t messed up  whitespace\n\n\n",
        r"\s",
        " dirty Strings with messed up whitespace\n"
    )]
    #[case(
        " dirty Strings  \t with  \t\t messed up  whitespace\n\n\n",
        r" ",
        " dirty Strings \t with \t\t messed up whitespace\n\n\n"
    )]
    fn test_squeeze(#[case] input: &str, #[case] pattern: Regex, #[case] expected: &str) {
        let stage = SqueezeStage {};

        let mut view = ScopedView::new(input);
        stage.substitute(&mut view);
        let result = view.to_string();

        assert_eq!(result, expected);
    }
}
