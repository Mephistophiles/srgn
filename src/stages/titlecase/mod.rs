use super::Stage;
use titlecase::titlecase;

/// Renders in titlecase.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
#[allow(clippy::module_name_repetitions)]
pub struct TitlecaseStage {}

impl Stage for TitlecaseStage {
    fn process(&self, input: &str) -> String {
        titlecase(input)
    }
}

#[cfg(test)]
mod tests {
    use rstest::rstest;

    use super::*;

    #[rstest]
    #[case("a dog", "A Dog")]
    #[case("ein überfall", "Ein Überfall")]
    #[case("miXeD caSe", "miXeD caSe")] // Hmmm... behavior of `titlecase` crate
    //
    #[case("a dog's life 🐕", "A Dog's Life 🐕")]
    //
    #[case("a dime a dozen", "A Dime a Dozen")]
    fn test_titlecasing(#[case] input: &str, #[case] expected: &str) {
        let result = TitlecaseStage::default().process(input);
        assert_eq!(result, expected);
    }
}
