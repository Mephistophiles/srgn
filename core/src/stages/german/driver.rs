use crate::stages::{
    german::{
        machine::{StateMachine, Transition},
        words::{Replace, Replacement, WordCasing},
    },
    tooling::StageResult,
    Stage,
};
use cached::proc_macro::cached;
use cached::SizedCache;
use common::lookup::binary_search_uneven;
use common::strings::is_compound_word;
use itertools::Itertools;
use log::{debug, trace};
use unicode_titlecase::StrTitleCase;

static VALID_GERMAN_WORDS: &str = include_str!(concat!(env!("OUT_DIR"), "/de.txt")); // Generated in `build.rs`.

/// German language stage, responsible for Umlauts and Eszett.
///
/// This stage is responsible for applying the following rules, **where applicable**:
/// - [*Umlauts*](https://en.wikipedia.org/wiki/Umlaut_(diacritic)): replace `ue`, `oe`,
///   `ae` with `√º`, `√∂`, `√§`, respectively,
/// - [*Eszett*](https://en.wikipedia.org/wiki/%C3%9F): replace `ss` with `√ü`.
///
/// Mechanisms are in place to uphold the following properties:
/// - both lower- and uppercase variants are handled correctly,
/// - compound words are handled correctly.
///
/// Towards this, this stage does *not* simply replace all occurrences, but performs
/// checks to ensure only valid replacements are made. The core of these checks is an
/// exhaustive word list. The better the word list, the better the results. As such, any
/// errors in processing could be the result of a faulty word list *or* faulty
/// algorithms.
///
/// # Example: A simple greeting, with Umlaut and Eszett
///
/// ```
/// use betterletter::{Stage, stages::GermanStage};
///
/// let result: String = GermanStage.substitute("Gruess Gott!").unwrap().into();
/// assert_eq!(result, "Gr√º√ü Gott!");
/// ```
///
/// # Example: A compound word
///
/// Note that this compound word is *not* part of the word list (that would be an
/// *elaborate* word list!), but is still handled, as its constituents are.
///
/// ```
/// use betterletter::{Stage, stages::GermanStage};
///
/// let result: String = GermanStage.substitute("Du Suesswassertagtraeumer!").unwrap().into();
/// assert_eq!(result, "Du S√º√üwassertagtr√§umer!");
/// ```
///
/// # Example: Words *validly* containing alternative Umlaut spelling
///
/// These spellings are *not* replaced, as they are valid words in their own right.
/// Naive implementations/translations (e.g.
/// [`tr`](https://en.wikipedia.org/wiki/Tr_(Unix))) would not handle this correctly.
///
/// ```
/// use betterletter::{Stage, stages::GermanStage};
///
/// for word in &[
///     // "ae"
///     "Aerodynamik",   // should not be "√Ñrodynamik"
///     "Israel",        // should not be "Isr√§l"
///     "Schufaeintrag", // should not be "Schuf√§intrag"
///     // "oe"
///     "Koeffizient",   // should not be "K√∂ffizient"
///     "Dominoeffekt",  // should not be "Domin√∂ffekt"
///     "Poet",          // should not be "P√∂t"
///     // "ue"
///     "Abenteuer",     // should not be "Abente√ºr"
///     "Mauer",         // should not be "Ma√ºr"
///     "Steuerung",     // should not be "Ste√ºrung"
/// ] {
///     let result: String = GermanStage.substitute(word).unwrap().into();
///     assert_eq!(result, word.to_string());
/// }
/// ```
///
/// Note that `ss`/`√ü` is not mentioned, as it is handled
/// [elsewhere](#example-words-with-valid-alternative-and-special-character-spellings).
///
/// # Example: Words with valid alternative *and* special character spellings
///
/// Some words are validly spelled with alternative Umlauts *and* special characters *in
/// the same position*, such as:
/// - [Ma**√ü**e](https://de.wiktionary.org/wiki/Ma%C3%9Fe): pertaining to measurements
/// - [Ma**ss**e](https://de.wiktionary.org/wiki/Masse): pertaining to mass/weight
///
/// So if a user inputs `Masse` (they can't spell `Ma√üe`, else they wouldn't have
/// reached for this crate in the first place), what do they mean? Such cases are
/// tricky, as there isn't an easy solution without reaching for full-blown
/// [NLP](https://en.wikipedia.org/wiki/Natural_language_processing) or ML, as the
/// word's context would be required. This stage is much too limited for that. A choice
/// has to be made:
///
/// - do not replace: keep alternative spelling, or
/// - replace: keep special character spelling.
///
/// This tool chooses the latter, as it seems [the least
/// astonishing](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) in the
/// context of this tool, whose entire point is to **make replacements if they're
/// valid**.
///
/// This is an issue mainly for Eszett (`√ü`), as for it, two valid spellings are much
/// more likely than for Umlauts.
///
/// ```
/// use betterletter::{Stage, stages::GermanStage};
///
/// for (input, output) in &[
///     ("Busse", "Bu√üe"), // busses / penance
///     ("Masse", "Ma√üe"), // mass / measurements
/// ] {
///     let result: String = GermanStage.substitute(input).unwrap().into();
///     assert_eq!(result, output.to_string());
/// }
/// ```
///
/// # Example: Upper- and mixed case
///
/// This stage can handle any case, but assumes **nouns are never lower case** (a pretty
/// mild assumption). The **first letter governs the case** of the entity (Umlaut,
/// Eszett or entire word) in question:
///
/// | Input | Example Umlaut/Eszett | Example word | Detected case |
/// | ----- | --------------------- | ------------ | ------------- |
/// | `xx`  | `ue`                  | `hello`      | lowercase     |
/// | `xX`  | `sS`                  | `hElLo`      | lowercase     |
/// | `Xx`  | `Ue`                  | `Hello`      | uppercase     |
/// | `XX`  | `SS`                  | `HELLooo`    | uppercase     |
///
/// The same principle then further applies to entire words, which is especially
/// noticeable for mixed-case ones. The word list is not going to contain mixed-case
/// words, so a decision has to be made: what case will candidates be checked against?
/// If whatever case was detected is not considered a valid word, the replacement is not
/// made. Example flows follow.
///
/// ## Subexample: mixed case, invalid word
///
/// The flow looks like:
///
/// `aEpFeL` ‚Üí lowercase Umlaut ‚Üí `√§pFeL` ‚Üí lowercase word ‚Üí squash ‚Üí `√§pfel` ‚Üí ‚ùå ‚Üí
/// output is `aEpFeL`
///
///
/// ```
/// use betterletter::{Stage, stages::GermanStage};
///
/// let result: String = GermanStage.substitute("aEpFeL").unwrap().into();
///
/// // Error: MiXeD CaSe noun without leading capital letter
/// assert_eq!(result, "aEpFeL");
/// ```
///
/// ## Subexample: mixed case, valid word
///
/// The flow looks like:
///
/// `AePfEl` ‚Üí uppercase Umlaut ‚Üí `√ÑPfEl` ‚Üí uppercase word ‚Üí squash ‚Üí `√Ñpfel` ‚Üí ‚úÖ ‚Üí
/// output is `√Ñpfel`
///
/// ```
/// use betterletter::{Stage, stages::GermanStage};
///
/// let result: String = GermanStage.substitute("AePfEl").unwrap().into();
///
/// // OK: MiXeD CaSe words nouns are okay, *if* starting with a capital letter
/// assert_eq!(result, "√ÑPfEl");
/// ```
///
/// ## Subexample: other cases
///
/// ```
/// use betterletter::{Stage, stages::GermanStage};
///
/// let f = |word: &str| -> String {GermanStage.substitute(word).unwrap().into()};
///
/// // OK: The normal case, adjective lowercase
/// assert_eq!(f("Voll suess!"), "Voll s√º√ü!");
///
/// // OK: Adjective uppercase (start of sentence)
/// assert_eq!(f("Suesses Eis!"), "S√º√ües Eis!");
///
/// // OK: Uppercased noun
/// assert_eq!(f("Aepfel"), "√Ñpfel");
///
/// // Error: Lowercased noun is *not* replaced, we are not a spell checker
/// assert_eq!(f("aepfel"), "aepfel");
///
/// // OK: SCREAMING CASE noun is okay though
/// assert_eq!(f("AEPFEL"), "√ÑPFEL");
///
/// // OK: SCREAMING CASE verb is okay as well
/// assert_eq!(f("SCHLIESSEN"), "SCHLIE·∫ûEN");
///
/// // OK: MiXeD CaSe verb: inserted special character is uppercase
/// assert_eq!(f("fUeLleN"), "f√úLleN");
///
/// // OK: MiXeD CaSe verb: inserted special character is lowercase
/// assert_eq!(f("FuElLEn"), "F√ºlLEn");
/// ```
///
/// ### Capital Eszett (·∫û)
///
/// Note the spelling of `SCHLIE·∫ûEN` containing `·∫û`, the [uppercase version of
/// `√ü`](https://www.wikidata.org/wiki/Q9693), part of [official spelling since
/// 2017](https://web.archive.org/web/20230206102049/https://www.rechtschreibrat.com/DOX/rfdr_PM_2017-06-29_Aktualisierung_Regelwerk.pdf).
/// It's the result of uppercasing `√ü` of `schlie√üen`. This does **not** follow Rust's
/// usual behavior, which is why it is specially mentioned here:
///
/// ```
/// let lc = "√ü";
/// let uc = "·∫û";
///
/// assert_eq!(lc.to_uppercase().to_string(), "SS");
///
/// // The other way around works though:
/// assert_eq!(uc.to_lowercase().to_string(), lc);
///
/// // Uppercase stays uppercase:
/// assert_eq!(uc.to_uppercase().to_string(), uc);
///
/// // Lowercase stays lowercase (as opposed to `ss`):
/// assert_eq!(lc.to_lowercase().to_string(), lc);
/// ```
///
/// The `SS` of `SCHLIESSEN` is detected as an uppercase Eszett, which is specifically
/// inserted. You might want to run additional processing if this is undesired.
///
/// # Example: Other bytes
///
/// This stage handles the German alphabet *only*, and will leave other input bytes
/// untouched. You get to keep your trailing newlines, emojis (also multi-[`char`] ones),
/// and everything else.
///
/// Of course, the input has to be valid UTF-8, as is ensured by its signature ([`str`]).
///
/// ```
/// use betterletter::{Stage, stages::GermanStage};
///
/// let result: String = GermanStage.substitute("\0Schoener    ‰Ω†Â•Ω Satz... üëãüèª\r\n\n").unwrap().into();
/// assert_eq!(result, "\0Sch√∂ner    ‰Ω†Â•Ω Satz... üëãüèª\r\n\n");
/// ```
///
/// # Performance
///
/// This stage is implemented as a [finite state
/// machine](https://en.wikipedia.org/wiki/Finite-state_machine), which means it runs in
/// linear time as well as constant space. It is therefore very fast and memory
/// efficient, requiring only a single pass over the input [`str`].
///
/// The underlying checks for valid words are implemented as a
/// [memoized](https://en.wikipedia.org/wiki/Memoization), recursive binary search.
/// While they're fast, other methods could be faster but weren't chosen for one or more
/// of these reasons:
///
/// - poor developer experience:
///    - [`clippy`](https://github.com/rust-lang/rust-clippy) would choke on them
///    - compilation times of 5 minutes and more (on fast hardware)
/// - large binary size:
///
///   A simple array of strings, `&[&str]`, adds two [`usize`] in terms of overhead **per
///   [`str`]** (tuple of `(pointer, length)`), which is 16 bytes on 64-bit systems and
///   therefore **longer than the average word** (which sits at around 15 bytes, give or
///   take). Seeing as there can be hundreds of thousands, if not millions of entries,
///   this quickly *doubles* the binary size for no good reason.
/// - not available statically, aka at compile time, aka incurring a runtime cost. This
///   crate's binary is optimized for start-up speed.
///
/// For more info, an overview of the methods tried
/// ([`phf`](https://crates.io/crates/phf) and more), and benchmarks, see [this
/// issue](https://github.com/alexpovel/betterletter-rs/issues/9).
#[derive(Debug, Clone, Copy)]
pub struct GermanStage;

impl Stage for GermanStage {
    fn substitute(&self, input: &str) -> StageResult {
        const INDICATOR: char = '\0';

        debug!("Working on input '{}'", input.escape_debug());

        let mut output = String::with_capacity(input.len());
        let mut machine = StateMachine::new();

        // The state machine, much like a missing trailing newline in a file, will
        // misbehave if the very last transition is not an 'external' one (the last word
        // won't be detected properly).
        for char in input.chars().chain(std::iter::once(INDICATOR)) {
            trace!(
                "Beginning processing of character '{}'",
                char.escape_debug()
            );

            let transition = machine.transition(char);

            trace!("Transition is '{:?}'", transition);

            match transition {
                Transition::External => {
                    output.push(char);
                    continue;
                }
                Transition::Entered | Transition::Internal => {
                    continue;
                }
                Transition::Exited => {
                    debug!("Exited machine: {:?}", machine);

                    let original = machine.current_word().content().to_owned();
                    let word =
                        find_valid_replacement(&original, machine.current_word().replacements())
                            .unwrap_or(original);

                    debug!("Processed word, appending to output: {:?}", &word);
                    output.push_str(&word);

                    // Add back the non-word character that caused the exit transition in the
                    // first place.
                    output.push(char);
                }
            }
        }

        let c = output.pop();
        debug_assert!(
            c == Some(INDICATOR),
            "Trailing indicator byte expected, but found '{c:?}'."
        );

        debug!("Final output string is '{}'", output.escape_debug());

        Ok(output.into())
    }
}

fn find_valid_replacement(word: &str, replacements: &[Replacement]) -> Option<String> {
    let replacement_combinations: Vec<Vec<Replacement>> = replacements
        .iter()
        .powerset()
        .map(|v| v.into_iter().copied().collect())
        .collect();

    debug!("Starting search for valid replacement for word '{}'", word);
    trace!(
        "All replacement combinations to try: {:?}",
        replacement_combinations
    );

    // By definition, the power set contains the empty set. There are two options for
    // handling it:
    // - not skipping: empty set is tried first, and if that word is valid, it is
    //   returned
    // - skipping: empty set is skipped, *some* replacements will take place; if none of
    //   them are valid, no replacements will take place
    //
    // Not skipping it means words like `Busse` will remain unchanged on first
    // iteration. Then, `Busse` will turn out to be valid already and will be returned .
    // Skipping it means `Bu√üe` is tried, which is *also* valid and returned, foregoing
    // `Busse`.
    debug_assert!(replacement_combinations
        .first()
        .map_or(true, std::vec::Vec::is_empty));

    for replacements in replacement_combinations.into_iter().skip(1) {
        let mut candidate = word.to_owned();
        candidate.apply_replacements(replacements);
        trace!(
            "Replaced candidate word, now is: '{}'. Starting validity check.",
            candidate
        );

        if is_valid(&candidate, &contained_in_global_word_list) {
            debug!("Candidate '{}' is valid, returning early", candidate);
            return Some(candidate);
        }

        trace!("Candidate '{}' is invalid, trying next one", candidate);
    }

    debug!("No valid replacement found, returning");
    None
}

fn contained_in_global_word_list(word: &str) -> bool {
    binary_search_uneven(word, VALID_GERMAN_WORDS, '\n')
}

// https://github.com/jaemk/cached/issues/135#issuecomment-1315911572
#[cached(
    type = "SizedCache<String, bool>",
    create = "{ SizedCache::with_size(1024) }",
    convert = r#"{ String::from(word) }"#
)]
fn is_valid(word: &str, predicate: &impl Fn(&str) -> bool) -> bool {
    trace!("Trying candidate '{}'", word);

    let casing = WordCasing::try_from(word);
    trace!("Casing of candidate is '{:?}'", casing);

    match casing {
        Ok(WordCasing::AllLowercase) => {
            // There is no further processing we can/want to do (or is there...
            // https://www.youtube.com/watch?v=HLRdruqQfRk).
            predicate(word)
        }
        Ok(WordCasing::AllUppercase) => {
            // Convert to something sensible before proceeding.
            let tc = word.to_titlecase_lower_rest();
            debug_assert!(
                WordCasing::try_from(tc.as_str()) == Ok(WordCasing::Titlecase),
                "Titlecased word, but isn't categorized correctly."
            );

            is_valid(&tc, predicate)
        }
        Ok(WordCasing::Mixed) => {
            // For MiXeD casing, the word's first character governs its further
            // treatment.
            match word.chars().next() {
                Some(c) if c.is_uppercase() => {
                    let tc = word.to_titlecase_lower_rest();
                    debug_assert!(
                        WordCasing::try_from(tc.as_str()) == Ok(WordCasing::Titlecase),
                        "Titlecased word, but isn't categorized correctly."
                    );

                    is_valid(&tc, predicate)
                }
                _ => is_valid(&word.to_lowercase(), predicate),
            }
        }
        Ok(WordCasing::Titlecase) => {
            // Regular nouns are normally titlecase, so see if they're found
            // immediately (e.g. "Haus").
            predicate(word)
                // Adjectives and verbs might be titlecased at the beginning of
                // sentences etc. (e.g. "Gut gemacht!" -> we need "gut").
                || is_valid(&word.to_lowercase(), predicate)
                // None of these worked: we might have a compound word. These are
                // *never* assumed to occur as anything but titlecase (e.g.
                // "Haus√ºberfall").
                || is_compound_word(word, predicate)
        }
        Err(_) => false, // Ran into some unexpected characters...
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use common::instrament;
    use itertools::Itertools;
    use rstest::rstest;

    #[test]
    fn test_words_are_sorted() {
        let original = VALID_GERMAN_WORDS.lines().collect_vec();

        let mut sorted = VALID_GERMAN_WORDS.lines().collect_vec();
        sorted.sort_unstable(); // see also: clippy::stable_sort_primitive

        assert_eq!(original, sorted.as_slice());
    }

    #[test]
    fn test_words_are_unique() {
        let original = VALID_GERMAN_WORDS.lines().collect_vec();

        let mut unique = VALID_GERMAN_WORDS.lines().collect_vec();
        unique.sort_unstable(); // see also: clippy::stable_sort_primitive
        unique.dedup();

        assert_eq!(original, unique.as_slice());
    }

    #[test]
    fn test_word_list_is_not_filtered() {
        assert!(
            VALID_GERMAN_WORDS.lines().any(str::is_ascii),
            concat!(
                "Looks like you're using a filtered word list containing only special characters.",
                " The current implementation relies on the full word list (also containing all non-Umlaut words)"
            )
        );
    }

    #[test]
    fn test_is_valid_on_empty_input() {
        assert!(!is_valid("", &contained_in_global_word_list));
    }

    instrament! {
        #[rstest]
        fn test_is_valid(
            #[values(
                "????",
                "\0",
                "\0D√ºbel",
                "\0D√ºbel\0",
                "ü§©D√ºbel",
                "ü§©D√ºbelü§ê",
                "üòé",
                "dr√∂ge",
                "D√ºBeL",
                "D√ºbel\0",
                "Duebel",
                "kindergarten",
                "Koeffizient",
                "kongruent",
                "K√ºbel",
                "Mauer",
                "MauerüòÇ",
                "Mauerd√ºbel",
                "Mauerd√ºbelk√ºbel",
                "Ma√ºr",
                "Ma√ºrd√ºbelk√ºbel",
                "messger√§t",
                "No\nway",
                "S√º√üwasserschwimmb√§der",
                "ŸÖÿ±ÿ≠ÿ®ÿß",
                "‰Ω†Â•Ω",
            )]
            word: String
        ) (|data: &TestIsValid| {
                insta::assert_yaml_snapshot!(data.to_string(), is_valid(&word, &contained_in_global_word_list));
            }
        )
    }

    instrament! {
        #[rstest]
        fn test_process(
            #[values(
                "\0Kuebel",
                "\0Duebel\0",
                "ü§©Duebel",
                "ü§©Duebelü§ê",
                "D√ºbel",
                "Abenteuer sind toll!",
                "Koeffizient",
                "kongruent",
                "Ich mag Aepfel, aber nicht Aerger.",
                "Ich mag AEPFEL!! üòç",
                "Wer mag Aepfel?!",
                "Was sind aepfel?",
                "Oel ist ein wichtiger Bestandteil von Oel.",
                "WARUM SCHLIESSEN WIR NICHT AB?",
                "Wir schliessen nicht ab.",
                "WiR sChLieSsEn ab!",
                "WiR sChLiesSEn vieLleEcHt aB.",
                "Suess!",
            )]
            word: String
        ) (|data: &TestProcess| {
                let input = word.clone();
                let result = GermanStage{}.substitute(&input).unwrap();
                insta::assert_yaml_snapshot!(data.to_string(), result.0);
            }
        )
    }
}
