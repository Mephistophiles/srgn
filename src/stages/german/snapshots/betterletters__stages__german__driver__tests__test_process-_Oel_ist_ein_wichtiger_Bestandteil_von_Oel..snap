---
source: core/src/stages/german/driver.rs
expression: result.0
info:
  word: Oel ist ein wichtiger Bestandteil von Oel.
---
Öl ist ein wichtiger Bestandteil von Öl.

